src/stores/AppStore.ts

```diff
@@ -929,30 +931,39 @@ export class AppStore {
         try {
             const ack = await this.backendService.importRegion(directory, file, type, frame.frameInfo.fileId);
             if (frame && ack.success && ack.regions) {
-                const regionMap = new Map<string, CARTA.IRegionInfo>(Object.entries(ack.regions));
-                const regionStyles = new Map<string, CARTA.IRegionStyle>(Object.entries(ack.regionStyles));
-                regionMap.forEach((regionInfo, regionIdString) => {
-                    const styleInfo = regionStyles.get(regionIdString);
-
-                    frame.regionSet.addExistingRegion(
-                        regionInfo.controlPoints as Point2D[],
-                        regionInfo.rotation,
-                        regionInfo.regionType,
-                        parseInt(regionIdString),
-                        styleInfo?.name,
-                        styleInfo?.color,
-                        styleInfo?.lineWidth,
-                        styleInfo?.dashList
-                    );
-                });
+                const regions = Object.entries(ack.regions);
+                const regionStyleMap = new Map<string, CARTA.IRegionStyle>(Object.entries(ack.regionStyles));
+                this.addRegionsInBatch(regions, regionStyleMap, 0);
             }
-            this.fileBrowserStore.hideFileBrowser();
         } catch (err) {
             console.error(err);
             AppToaster.show(ErrorToast(err));
         }
     };
 
+    addRegionsInBatch = (regions: [string, CARTA.IRegionInfo][], regionStyleMap: Map<string, CARTA.IRegionStyle>, count: number) => {
+        if (!regions || !regionStyleMap || !isFinite(count)) {
+            return;
+        }
+
+        const frame = this.activeFrame;
+        const batchEnd = Math.min(count + IMPORT_REGION_BATCH_SIZE, regions.length);
+        for (; count < batchEnd; count++) {
+            const [regionIdString, regionInfo] = regions[count];
+            const styleInfo = regionStyleMap.get(regionIdString);
+            frame.regionSet.addExistingRegion(regionInfo.controlPoints as Point2D[], regionInfo.rotation, regionInfo.regionType, parseInt(regionIdString), styleInfo?.name, styleInfo?.color, styleInfo?.lineWidth, styleInfo?.dashList);
+        }
+        this.fileBrowserStore.updateLoadingState(count / regions.length, count, regions.length);
+
+        if (count < regions.length) {
+            setTimeout(this.addRegionsInBatch, 0, regions, regionStyleMap, count);
```

Comment:
I had no idea you could pass additional arguments to `setTimeout` like this :+1: